lightning/src/util/ser.rs

```diff
@@ -67,6 +68,46 @@ impl Writer for VecWriter {
 	}
 }
 
+pub(crate) struct LengthCalculatingWriter(pub usize);
+impl Writer for LengthCalculatingWriter {
+	#[inline]
+	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0 += buf.len();
+		Ok(())
+	}
+	#[inline]
+	fn size_hint(&mut self, _size: usize) {}
+}
+
+/// Essentially std::io::Take but a bit simpler and exposing the amount read at the end, cause we
+/// may need to skip ahead that much at the end.
+pub(crate) struct FixedLengthReader<R: Read> {
+	pub read: R,
+	pub read_len: u64,
+	pub max_len: u64,
+}
+impl<R: Read> FixedLengthReader<R> {
+	pub fn eat_remaining(&mut self) -> Result<(), ::std::io::Error> {
+		while self.read_len != self.max_len {
+			debug_assert!(self.read_len < self.max_len);
+			let mut buf = [0; 1024];
+			let readsz = cmp::min(1024, self.max_len - self.read_len) as usize;
+			self.read_exact(&mut buf[0..readsz])?;
+		}
+		Ok(())
+	}
```

Comment:
Instead of rolling our own implementation, we can use `std::io` to simplify this as:

```rust
pub fn eat_remaining(self) -> Result<(), ::std::io::Error> {
    let mut remaining = self.read.take(self.max_len - self.read_len);
    let bytes_discarded = std::io::copy(&mut remaining, &mut std::io::sink())?;
    // ...
}
```

Reply:
Good idea. Don't even have to create the take, we can just use self (though this did turn up this stream being super awkward, so I rewrote read() to be a bit more friendly to other std::io types.